sklearn/ensemble/_forest.py

```diff
@@ -426,38 +494,111 @@ def fit(self, X, y, sample_weight=None):
 
         return self
 
+    def _compute_impurity_importances(self):
+        """Compute the impurity-based feature importances.
+
+        Returns
+        -------
+        importances : ndarray of shape (n_features,)
+            The impurity-based feature importances.
+        """
+        parallel_args = {
+            **_joblib_parallel_args(prefer="threads"),
+            "n_jobs": self.n_jobs
+        }
+        all_importances = Parallel(**parallel_args)(
+            delayed(getattr)(tree, 'feature_importances_')
+            for tree in self.estimators_ if tree.tree_.node_count > 1
+        )
+        if not all_importances:
+            return np.zeros(
+                shape=(self.n_features_in_, 1), dtype=np.float64
+            )
+        return np.transpose(all_importances)
+
     @abstractmethod
-    def _set_oob_score_and_attributes(self, X, y):
+    def _set_oob_score_and_attributes(self, X, y, sample_weight):
         """Compute and set the OOB score and attributes.
 
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
             The data matrix.
+
         y : ndarray of shape (n_samples, n_outputs)
             The target matrix.
+
+        sample_weight : ndarray of shape (n_samples,)
+            Sample weights.
         """
 
-    def _compute_oob_predictions(self, X, y):
-        """Compute and set the OOB score.
+    def _compute_oob_importances(self, X, y, sample_weight):
+        """Compute importances by permuting features using OOB samples.
 
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
             The data matrix.
+
         y : ndarray of shape (n_samples, n_outputs)
             The target matrix.
 
+        sample_weight : ndarray of shape (n_samples,)
+            Sample weights.
+
+        Returns
+        -------
+        oob_importances : ndarray of shape (n_features, n_estimators)
+            The permutation feature importance compuring on OOB.
+        """
+        X = check_array(X, dtype=DTYPE, accept_sparse="csr")
+        random_state = check_random_state(self.random_state)
+
+        n_samples = X.shape[0]
+        n_samples_bootstrap = _get_n_samples_bootstrap(
+            n_samples,
+            self.max_samples,
+        )
+
+        with config_context(assume_finite=True):
+            # avoid redundant checking performed on X in the permutation
+            # importance function.
+            oob_importances = np.transpose(Parallel(n_jobs=self.n_jobs)(
+                delayed(_permutation_importances_oob)(
+                    estimator,
+                    X,
+                    y,
+                    sample_weight,
+                    n_samples,
+                    n_samples_bootstrap,
+                    random_state,
+                )
+                for estimator in self.estimators_
+            ))
+
+        return oob_importances
```

Comment:
Should we prefer threads here?

```suggestion
        with config_context(assume_finite=True):
            # avoid redundant checking performed on X in the permutation
            # importance function.
            parallel_args = {
                **_joblib_parallel_args(prefer="threads"),
                "n_jobs": self.n_jobs
            }
            oob_importances = Parallel(**parallel_args)(
                delayed(_permutation_importances_oob)(
                    estimator,
                    X,
                    y,
                    sample_weight,
                    n_samples,
                    n_samples_bootstrap,
                    random_state,
                )
                for estimator in self.estimators_
            ))
        oob_importances = np.transpose(oob_importances)
        return oob_importances
```

Reply:
I was under the impression that I did look at it before and the threading was not the best choice (not sure that we are releasing the GIL enough).